sympy/ntheory/primetest.py

```diff
@@ -110,6 +110,32 @@ def mr(n, bases):
     return True
 
 
+
+def is_euler_pseudoprime(n, b):
+    from sympy.ntheory.factor_ import trailing
+    """Tests whether n is prime or an Euler pseudoprime to base b."""
+    if not mr(n, [b]):
+        return False
+    n = as_int(n)
+    r = n-1
+    r = as_int(r)
+    s = trailing(r)
+    s = as_int(s)
+    r = r //(2**s)
+    c = pow(b, r, n)
+    if c == 1:
+        return True
+    
```

Comment:
It looks like you are not that familiar with Python requirements or else are using an editor that is inserting indentation for you. The quality tests will not pass if there is trailing whitespace on a line (as there is on line 128).

I recommend the following code for what you have written:

```
def is_euler_pseudoprime(n, b):
    """Return True if n is prime or an Euler pseudoprime to base b, else False."""
    from sympy.ntheory.factor_ import trailing

    if not mr(n, [b]):
        return False

    n = as_int(n)
    r = n - 1
    c = pow(b, r >> trailing(r), n)
    if c == 1:
        return True

    while True:
        if c == n - 1:
            return True
        c = pow(c, 2, n)
        if c == 1:
            return False

```

Reply:
Also, since there are 4 return statements, you should add 4 tests: one pair of n,b that return from each of those `return` lines. The tests should go in `test_primetest.py` in the tests folder of ntheory.

But...before you do more. What will the ` Solovay-Strassen` test do that is an improvement over what is already in SymPy? I ask because my quick read of the wiki on the method says that the Miller-Rabin test generally is preferred: "It has been largely superseded by the Baillie-PSW primality test and the Miller–Rabin primality test, but has great historical importance" [ref](https://en.wikipedia.org/wiki/Solovay%E2%80%93Strassen_primality_test).

Reply:
Sir,
I agree with your point on M-R being better than Solovay-Strassen but,
M-R is a probabilistic test which is based on the assumption of the Extended Riemann Hypothesis which no one has yet proved to be correct.
Also, all prime tests are probabilistic in sympy.My main idea is to add a deterministic prime test like AKS and/or APR-CL with runtime less than miller rabin test !!!!
https://en.wikipedia.org/wiki/Adleman%E2%80%93Pomerance%E2%80%93Rumely_primality_test

For these I need Euler pseudoprimes